alacritty_terminal/src/term/mod.rs

```diff
@@ -153,57 +199,172 @@ impl SizeInfo {
+        *self = Self::undamaged(self.line, num_cols);
     }
 
-    /// Calculate padding to spread it evenly around the terminal content.
     #[inline]
-    fn dynamic_padding(padding: f32, dimension: f32, cell_dimension: f32) -> f32 {
-        padding + ((dimension - 2. * padding) % cell_dimension) / 2.
+    pub fn expand(&mut self, left: usize, right: usize) {
+        self.left = cmp::min(self.left, left);
+        self.right = cmp::max(self.right, right);
+    }
+
+    #[inline]
+    pub fn is_damaged(&self) -> bool {
+        self.left <= self.right
     }
 }
 
-impl Dimensions for SizeInfo {
+/// Terminal damage information collected from the latest [`Term::reset_damage`] call.
+#[derive(Debug)]
+pub enum TermDamage<'a> {
+    /// The entire terminal is damaged.
+    Full,
+
+    /// Iterator over damaged lines in terminal.
+    Partial(TermDamageIterator<'a>),
+}
+
+/// Iterator over the terminal's damaged lines.
+#[derive(Clone, Debug)]
+pub struct TermDamageIterator<'a> {
+    line_damage: std::slice::Iter<'a, LineDamageBounds>,
+}
+
+impl<'a> TermDamageIterator<'a> {
+    fn new(line_damage: &'a [LineDamageBounds]) -> Self {
+        Self { line_damage: line_damage.iter() }
+    }
+}
+
+impl<'a> Iterator for TermDamageIterator<'a> {
+    type Item = LineDamageBounds;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.line_damage.find(|line| line.is_damaged()).copied()
+    }
+}
+
+/// State of the terminal damaged information.
+struct TermDamageState {
+    /// Hint whether terminal should be damaged entirely regardless of the actual damage changes.
+    is_fully_damaged: bool,
+
+    /// Information about damage on terminal lines.
+    lines: Vec<LineDamageBounds>,
+
+    /// Old terminal cursor point.
+    last_cursor: Point,
+
+    /// Old selection range.
+    last_selection: Option<SelectionRange>,
+}
+
+impl TermDamageState {
+    fn new(num_cols: usize, num_lines: usize) -> Self {
+        let lines =
+            (0..num_lines).map(|line| LineDamageBounds::undamaged(line, num_cols)).collect();
+
+        Self { is_fully_damaged: true, lines, last_cursor: Point::default(), last_selection: None }
+    }
+
     #[inline]
-    fn columns(&self) -> usize {
-        self.columns
+    fn resize(&mut self, num_cols: usize, num_lines: usize) {
+        // Reset point, so old cursor won't end up outside of the viewport.
+        self.last_cursor = Point::default();
+        self.last_selection = None;
+        self.is_fully_damaged = true;
+
+        self.lines.clear();
+        self.lines.reserve(num_lines);
+        for line in 0..num_lines {
+            self.lines.push(LineDamageBounds::undamaged(line, num_cols));
+        }
     }
 
     #[inline]
-    fn screen_lines(&self) -> usize {
-        self.screen_lines
+    fn damage_point(&mut self, point: Point) {
+        let line = point.line.0 as usize;
+        let column = point.column.0 as usize;
+        self.damage_line(line, column, column);
     }
```

Comment:
No, we don't damage offscreen points, however I do ensure that point is in viewport when calling to this function. The terminal point just can't have line and col as negative, since it won't make any sense. However Vi cursor could, and it converts into viewport coordinates before calling it.

Yeah, the comment should help.